In [1]:
import pandas as pd
import numpy as np

In [2]:
fdata = pd.read_csv('flight_data.csv')
fdata.head()

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   
3  2013      1    1     544.0             545       -1.0    1004.0   
4  2013      1    1     554.0             600       -6.0     812.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   
3            1022      -18.0      B6     725  N804JB    JFK  BQN     183.0   
4             837      -25.0      DL     461  N668DN    LGA  ATL     116.0   

   distance  hour  minute         time_hour  
0      1400     5      15  01-01-2013 05:00  
1      1416     5      29  01-01-2013 05:00  
2      1089     5      40  01-01-2013 05:00  
3      1576     5      45  01-01-2013 05:00  
4       762     6       0  01-01-2013 06:00

In [3]:
fdata.dtypes

year                int64
month               int64
day                 int64
dep_time          float64
sched_dep_time      int64
dep_delay         float64
arr_time          float64
sched_arr_time      int64
arr_delay         float64
carrier            object
flight              int64
tailnum            object
origin             object
dest               object
air_time          float64
distance            int64
hour                int64
minute              int64
time_hour          object
dtype: object

In [4]:
fdata.count

<bound method DataFrame.count of         year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     542.0             540        2.0     923.0   
3       2013      1    1     544.0             545       -1.0    1004.0   
4       2013      1    1     554.0             600       -6.0     812.0   
5       2013      1    1     554.0             558       -4.0     740.0   
6       2013      1    1     555.0             600       -5.0     913.0   
7       2013      1    1     557.0             600       -3.0     709.0   
8       2013      1    1     557.0             600       -3.0     838.0   
9       2013      1    1     558.0             600       -2.0     753.0   
10      2013      1    1     558.0             600       -2.0     849.0   
11      2013      1    1     558.0             600       -2.0     8

# Note :
1. If a column contains numbers and NaNs pandas will default to float64, in case your missing value has a decimal.
2. This means that our flight_dataset may contain null values and we need to remove them 

# Check Null Values in Dataset  : 



Total Number of rows in Dataset :

In [8]:
fdata.isnull().count()

year              336776
month             336776
day               336776
dep_time          336776
sched_dep_time    336776
dep_delay         336776
arr_time          336776
sched_arr_time    336776
arr_delay         336776
carrier           336776
flight            336776
tailnum           336776
origin            336776
dest              336776
air_time          336776
distance          336776
hour              336776
minute            336776
time_hour         336776
dtype: int64

Does our data have null values ?? 

In [9]:
fdata.isnull().values.any()

True

Total Number of Null Values :

In [10]:
fdata.isnull().sum()

year                 0
month                0
day                  0
dep_time          8255
sched_dep_time       0
dep_delay         8255
arr_time          8713
sched_arr_time       0
arr_delay         9430
carrier              0
flight               0
tailnum           2512
origin               0
dest                 0
air_time          9430
distance             0
hour                 0
minute               0
time_hour            0
dtype: int64

# Analysis of Null Values:
1. All sched_dep_time & sched_arr_time of flights are available but for some flights data for actual dep_time & arr_time of flights is NOT available.
2. count of arr_time is more compared to dep_time, which means these flights were departed but dep_time data is missing. -- Need to figure out how we can calculate this?
3. Rest all flights whose arr_time & dep_time is missing we can ignore (drop) this data rows.


In [11]:
fdata['arr_time'].isnull().sum() - fdata['dep_time'].isnull().sum()

458

* There are 458 Flights for which arrival time is NOT available but departure time is available. These are valid number of flights where dep_time data is missing. -- Keep these rows


Next Find Total rows in which both dep_time & arr_time are Null (Missing)

In [12]:
len(fdata[fdata['arr_time'].isnull() & fdata['dep_time'].isnull()])

8255

* We can drop above rows 

In [16]:
len(fdata)

336776

In [20]:
fdata.dropna(subset=['dep_time', 'arr_time'], how='all',inplace=True)

In [21]:
fdata.shape


(328521, 19)

#  Exploration ideas : 
# 1. <U> Departure Delays 

In [25]:
fdata['dep_delay'].max()

1301.0

In [26]:
fdata[fdata['dep_delay']==fdata['dep_delay']].max()


year                          2013
month                           12
day                             31
dep_time                      2400
sched_dep_time                2359
dep_delay                     1301
arr_time                      2400
sched_arr_time                2359
arr_delay                     1272
carrier                         YV
flight                        8500
tailnum                     N9EAMQ
origin                         LGA
dest                           XNA
air_time                       695
distance                      4983
hour                            23
minute                          59
time_hour         31-12-2013 23:00
dtype: object

# Convert time into HH:MM